cmd/erasure.go

```diff
@@ -272,38 +274,58 @@ func (er erasureObjects) LocalStorageInfo(ctx context.Context, metrics bool) Sto
 // - Non-scanning disks
 // - Non-healing disks
 // - Healing disks (if inclHealing is true)
-func (er erasureObjects) getOnlineDisksWithHealingAndInfo(inclHealing bool) (newDisks []StorageAPI, newInfos []DiskInfo, healing bool) {
-	var wg sync.WaitGroup
+func (er *erasureObjects) getOnlineDisksWithHealingAndInfo(inclHealing bool) (newDisks []StorageAPI, newInfos []DiskInfo, healing bool) {
 	disks := er.getDisks()
-	infos := make([]DiskInfo, len(disks))
-	r := rand.New(rand.NewSource(time.Now().UnixNano()))
-	for _, i := range r.Perm(len(disks)) {
-		i := i
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
 
-			disk := disks[i]
-			if disk == nil {
-				infos[i].Error = errDiskNotFound.Error()
-				return
-			}
+	infos := func() []DiskInfo {
+		var wg sync.WaitGroup
+		dinfos := make([]DiskInfo, len(disks))
+		r := rand.New(rand.NewSource(time.Now().UnixNano()))
+		for _, i := range r.Perm(len(disks)) {
+			i := i
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
 
-			di, err := disk.DiskInfo(context.Background(), DiskInfoOptions{})
-			infos[i] = di
-			if err != nil {
-				// - Do not consume disks which are not reachable
-				//   unformatted or simply not accessible for some reason.
-				infos[i].Error = err.Error()
+				disk := disks[i]
+				if disk == nil {
+					dinfos[i].Error = errDiskNotFound.Error()
+					return
+				}
+
+				di, err := disk.DiskInfo(context.Background(), DiskInfoOptions{})
+				dinfos[i] = di
+				if err != nil {
+					// - Do not consume disks which are not reachable
+					//   unformatted or simply not accessible for some reason.
+					dinfos[i].Error = err.Error()
+				}
+			}()
+		}
+		wg.Wait()
+		return dinfos
+	}
+
+	var dinfos []DiskInfo
+	if inclHealing {
+		// List/Walk calls ask for all disks including the ones under healing
+		// pick DiskInfo() for those disks as well.
+		er.onlineDisksHealingInfoCache.Once.Do(func() {
+			er.onlineDisksHealingInfoCache.TTL = time.Second
+			er.onlineDisksHealingInfoCache.Update = func() (interface{}, error) {
+				return infos(), nil
```

Comment:
`infos` closure binds to a point in time value of `er.getDisks()` via the variable `disks`. This will be called every time `er.onlineDisksHealingInfoCache` requires an update. 
We should not rely on `disks` and instead call `er.getDisks()` inside the closure `infos`.